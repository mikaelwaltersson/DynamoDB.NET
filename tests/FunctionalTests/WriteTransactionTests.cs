using DynamoDB.Net.Tests.FunctionalTests.Common;
using DynamoDB.Net.Tests.FunctionalTests.Common.Models;
using static DynamoDB.Net.DynamoDBExpressions;

namespace DynamoDB.Net.Tests.FunctionalTests.TestModels;

public class WriteTransactionTests : FunctionalTestSuiteBase
{
    [SeedItem]
    readonly User user = 
        new()
        { 
            UserId = new("4f7e4ea9-c642-43da-9c8b-169dc6a03824"), 
            Username = "johndoe", FullName = "John Doe",
            UserStatus = UserStatus.Active
        };
    
    [SeedItem]
    readonly Account account1 = 
        new()
        { 
            UserId = new("4f7e4ea9-c642-43da-9c8b-169dc6a03824"), 
            AccountName = "Savings Account", 
            AccountNumber = "123001-13330001", DateOpened = new(2024, 01, 20), 
            Balance = 2600.0m,
            AccountStatus = AccountStatus.Active
        };
    
    [SeedItem]
    readonly Account account2 = 
        new() 
        { 
            UserId = new("4f7e4ea9-c642-43da-9c8b-169dc6a03824"), 
            AccountName = "Transaction Account", 
            AccountNumber = "123001-13330002", DateOpened = new(2024, 01, 20), 
            Balance = 400.0m,
            AccountStatus = AccountStatus.Active
        };

    [SeedItem]
    readonly Account account3 = 
        new() 
        { 
            UserId = new("4f7e4ea9-c642-43da-9c8b-169dc6a03824"), 
            AccountName = "Transaction Account", 
            AccountNumber = "123001-13330003", DateOpened = new(2024, 01, 27), 
            Balance = 0.0m,
            AccountStatus = AccountStatus.Active
        };

    [SeedItem]
    readonly Transaction transaction1 = 
        new()
        { 
            TransactionId = new("70c0712e-276e-43d1-9dc4-db776cf24590"), 
            FromAccountNumber = "123001-13330001", ToAccountNumber = "123001-13330002",
            TransactionDate = new(2024, 01, 20), TimeOfDay = new(15, 23, 01), 
            Amount = 400m, TransactionType = TransactionType.Transfer 
        };
    

    [Fact]
    public async Task CanApplyMultipleOperationsInSingleWriteTransaction()
    {
        var newTransaction = 
            new Transaction()
            {
                TransactionId = new("6f708458-1e2a-4393-b69c-22b03455f5b9"), 
                FromAccountNumber = "123001-13330001", ToAccountNumber = "123001-13330003",
                TransactionDate = new(2024, 01, 27), TimeOfDay = new(10, 27, 26), 
                Amount = 300, TransactionType = TransactionType.Transfer 
            };

        await DynamoDBClient
            .BeginWriteTransaction()
            .ConditionCheck(
                user, 
                condition: () => user.UserStatus != UserStatus.Deleted)
            .Update(
                account1, 
                update: () => Set(account1.Balance, 2700m), 
                condition: () => account1.Balance == 2600m && account1.AccountStatus == AccountStatus.Active)
            .Update(
                account2, 
                update: () => 
                    Set(account2.Balance, 0m) & 
                    Set(account2.AccountStatus, AccountStatus.Terminated) &
                    Set(account2.Notes, 
                        ListAppend(
                            IfNotExists(account2.Notes, new()),
                            new[] { "Account has been terminated and transactions" })),
                condition: () => account2.Balance == 400m && account2.AccountStatus == AccountStatus.Active)
            .Update(
                account3, 
                update: () => Set(account3.Balance, 300m),
                condition: () => account3.Balance == 0m && account3.AccountStatus == AccountStatus.Active)
            .Delete(
                transaction1,
                condition: () => AttributeExists(transaction1.TransactionId))
            .Put(
                newTransaction,
                condition: () => AttributeNotExists(newTransaction.TransactionId))
            .CommitAsync();

        var storedItems = (await GetAllStoredRawItemsAsync()).ToSnapshotFriendlyObject();

        Snapshot.Match(new { storedItems });
    }    
}
