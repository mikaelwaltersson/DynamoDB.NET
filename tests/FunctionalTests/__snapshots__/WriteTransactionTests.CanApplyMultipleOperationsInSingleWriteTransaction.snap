{
  "storedItems": [
    {
      "Table": "accounts",
      "Items": [
        {
          "accountStatus": {
            "S": "terminated"
          },
          "dateOpened": {
            "S": "2024-01-20"
          },
          "notes": {
            "L": [
              {
                "S": "Account has been terminated and transactions"
              }
            ]
          },
          "balance": {
            "N": "0"
          },
          "accountName": {
            "S": "Transaction Account"
          },
          "$version": {
            "N": "1"
          },
          "accountNumber": {
            "S": "123001-13330002"
          },
          "userId": {
            "S": "4f7e4ea9-c642-43da-9c8b-169dc6a03824"
          },
          "$type": {
            "S": "DynamoDB.Net.Tests.FunctionalTests.Common.Models.Account, DynamoDB.Net.Tests, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null"
          }
        },
        {
          "accountStatus": {
            "S": "active"
          },
          "dateOpened": {
            "S": "2024-01-20"
          },
          "balance": {
            "N": "2700"
          },
          "accountName": {
            "S": "Savings Account"
          },
          "$version": {
            "N": "1"
          },
          "accountNumber": {
            "S": "123001-13330001"
          },
          "userId": {
            "S": "4f7e4ea9-c642-43da-9c8b-169dc6a03824"
          },
          "$type": {
            "S": "DynamoDB.Net.Tests.FunctionalTests.Common.Models.Account, DynamoDB.Net.Tests, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null"
          }
        },
        {
          "accountStatus": {
            "S": "active"
          },
          "dateOpened": {
            "S": "2024-01-27"
          },
          "balance": {
            "N": "300"
          },
          "accountName": {
            "S": "Transaction Account"
          },
          "$version": {
            "N": "1"
          },
          "accountNumber": {
            "S": "123001-13330003"
          },
          "userId": {
            "S": "4f7e4ea9-c642-43da-9c8b-169dc6a03824"
          },
          "$type": {
            "S": "DynamoDB.Net.Tests.FunctionalTests.Common.Models.Account, DynamoDB.Net.Tests, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null"
          }
        }
      ]
    },
    {
      "Table": "transactions",
      "Items": [
        {
          "transactionType": {
            "S": "transfer"
          },
          "fromAccountNumber": {
            "S": "123001-13330001"
          },
          "amount": {
            "N": "300"
          },
          "toAccountNumber": {
            "S": "123001-13330003"
          },
          "$version": {
            "N": "1"
          },
          "transactionDate": {
            "S": "2024-01-27"
          },
          "transactionId": {
            "S": "6f708458-1e2a-4393-b69c-22b03455f5b9"
          },
          "timeOfDay": {
            "S": "10:27:26.0000000"
          }
        }
      ]
    },
    {
      "Table": "users",
      "Items": [
        {
          "fullName": {
            "S": "John Doe"
          },
          "userStatus": {
            "S": "active"
          },
          "userId": {
            "S": "4f7e4ea9-c642-43da-9c8b-169dc6a03824"
          },
          "$type": {
            "S": "DynamoDB.Net.Tests.FunctionalTests.Common.Models.User, DynamoDB.Net.Tests, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null"
          },
          "username": {
            "S": "johndoe"
          }
        }
      ]
    }
  ]
}
